Safe `Camera::unproject_position()`

lawnjelly opened the pull request:
* Provides a `safe_unproject_position()` function that returns failure and prevents NaNs
* Wraps the new function from the old bound `unproject_position()` function
* Adds unit test

A problem occurs when the test position is on (or close to) the focal plane of a perspective camera matrix. Under these conditions, the perspective divide is 0, and the unprojected position is undefined (potentially infinite from the viewport centre).

This PR provides a safe version of unproject that detects this condition and returns false, and attempts to recover and provide a sensible result, while printing a warning once (in DEV builds).

We wrap this from the existing `unproject_position()` routine, providing more robust behaviour from the existing bound function, and preventing NaNs, while still returning a result that should process normally and give expected results.

4.x version of #96032
Fixes #96028 (for 4.x)

Bugsquad edit:
Fixes #121686


## Notes
* Prior to the `ERR_FAIL_COND`, `unproject_position()` had a major bug whereby it would return NaNs when fed some input data.
* We also deal with test points that are close to the focal plane with an epsilon. This is to prevent near infinite return values and math exceptions in calling code in rare situations.

## Orthographic projections
Works fine in standard orthographic projections (via `Projection::set_orthogonal()`) where w is a constant 1.0, and the fallback block doesn't run.

There is an exception though: where a user bypasses native camera and creates a custom oblique or sheared projection matrix, they could still end up with `p.d` near zero.

However, even in this case, the PR is still correct. The existing `unproject` function would result in NaN, and this safe version will prevent this problem.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

lawnjelly committed:
Add `Camera::safe_unproject_position()`

* Provides a safe unproject function that returns failure, and wraps it from the old
* Adds unit test

RedMser commented on scene/3d/camera_3d.cpp:
```diff
@@ -496,27 +496,76 @@ Vector<Vector3> Camera3D::get_near_plane_points() const {
-Point2 Camera3D::unproject_position(const Vector3 &p_pos) const {
-	ERR_FAIL_COND_V_MSG(!is_inside_tree(), Vector2(), "Camera is not inside scene.");
+bool Camera3D::safe_unproject_position(const Vector3 &p_pos, Point2 &r_result) const {
+	ERR_FAIL_COND_V_MSG(!is_inside_tree(), false, "Camera is not inside scene.");
```
Returning `false` in error case will also cause the `WARN_PRINT_ONCE` to print, which results in a confusing message.

lawnjelly replied:
Ah good point! I'll fix this. :+1:

belzecue commented:
Would love this fix making it into 4.x.  I remember seeing this message back in the day.  Just hit the issue again today.  Had a camera3D in a scene, everything fine, then all of a sudden getting spammed with " ERROR: scene/3d/camera_3d.cpp:495 - Condition "p.d == 0" is true. Returning: Point2()" every time I click on an object in the scene.  Reloaded scene from scratch, still happening.  Only deleting and re-adding the Camera stopped it.  No idea what set it off.

UPDATE: So everything's fine until I do an "align transform with view" on the Camera3D.  As soon as I do that, from any perspective in the editor, it triggers the message spam when I click on objects.

UPDATE2: Deleting the project's .godot folder fixes it.

lawnjelly commented:
Rebased to current master.

Two slight improvements (which I will port to 3.x):

* when 3D normal is 0,0, I now return the *centre* of the 2D viewport (rather than 0, 0, which is the top left)
* the case when the 3d point is up (+) needed the viewport direction to be flipped (-y) because 2d has y in opposite polarity
* added a unit test (this might need checking, because I'm a newbie at writing Godot tests)

clayjohn reviewed (approved):
Looks good! Reviewed in the core meeting

Calinou reviewed (approved):
Tested locally, it works as expected. Code looks good to me.

However, this can lead to a slight change in behavior when you use `unproject_position()` on a camera's own position. This snippet no longer prints an error, but it prints a different value now:

```gdscript
@tool
extends Node3D

func _ready() -> void:
	var t = $qs.unproject_position(Vector3($qs.position))
	print(t)
```

In `master`, `t` was `Vector2(0, 0)` while it's `Vector2(576, 324)` now (half of the default viewport size, `Vector2(1152, 648)`).

This new value is ultimately more logical, but it should be called out in the migration guide.

lawnjelly commented:
> In master, t was Vector2(0, 0) while it's Vector2(576, 324) now (half of the default viewport size, Vector2(1152, 648)).
> This new value is ultimately more logical, but it should be called out in the migration guide.

The old value was just incorrect, due to the bug.

Could be mentioned, but I'm not sure it's worth worrying about too much, this is usually an edge case rather than something users would be creating normally in their games.

Repiteo commented on scene/3d/camera_3d.cpp:
```diff
@@ -479,27 +479,82 @@ Vector<Vector3> Camera3D::get_near_plane_points() const {
+
+		// The viewport size here is irrelevant, we just want a high number
+		// (representing infinity) but not actually close to infinity to prevent
+		// knock on bugs if later maths later does something with these values.
```
```suggestion
		// knock on bugs if later math later does something with these values.
```

lawnjelly replied:
UK / US English difference.
You say potato, I say potato. :grin:

Repiteo reviewed (dismissed):
Needs rebase to account for new codespell version

Repiteo commented:
Thanks!
